src/signing/petition_types/petition_entity.rs

```diff
@@ -0,0 +1,557 @@
+            PetitionFactors::new_override(matrix.override_factors),
+        )
+    }
+
+    pub fn new_unsecurified(
+        intent_hash: IntentHash,
+        entity: AddressOfAccountOrPersona,
+        instance: HierarchicalDeterministicFactorInstance,
+    ) -> Self {
+        Self::new(
+            intent_hash,
+            entity,
+            PetitionFactors::new_unsecurified(instance),
+            None,
+        )
+    }
+
+    /// Returns `true` signatures requirement has been fulfilled, either by
+    /// override factors or by threshold factors
+    pub fn has_signatures_requirement_been_fulfilled(&self) -> bool {
+        self.status() == PetitionFactorsStatus::Finished(PetitionFactorsStatusFinished::Success)
+    }
+
+    fn union_of<F, T>(&self, map: F) -> IndexSet<T>
+    where
+        T: Eq + std::hash::Hash + Clone,
+        F: Fn(&PetitionFactors) -> IndexSet<T>,
+    {
+        self.both(
+            |l| map(l),
+            |t, o| {
+                t.unwrap_or_default()
+                    .union(&o.unwrap_or_default())
+                    .cloned()
+                    .collect::<IndexSet<T>>()
+            },
+        )
+    }
+
+    pub fn all_factor_instances(&self) -> IndexSet<OwnedFactorInstance> {
+        self.union_of(|l| l.factor_instances())
+            .into_iter()
+            .map(|f| OwnedFactorInstance::owned_factor_instance(self.entity.clone(), f.clone()))
+            .collect::<IndexSet<_>>()
+    }
+
+    pub fn all_skipped_factor_instance(&self) -> IndexSet<HierarchicalDeterministicFactorInstance> {
+        self.union_of(|f| f.all_skipped())
+    }
+
+    pub fn all_skipped_factor_sources(&self) -> IndexSet<FactorSourceIDFromHash> {
+        self.all_skipped_factor_instance()
+            .into_iter()
+            .map(|f| f.factor_source_id)
+            .collect::<IndexSet<_>>()
+    }
+
+    pub fn all_signatures(&self) -> IndexSet<HDSignature> {
+        self.union_of(|f| f.all_signatures())
+    }
+
+    fn with_list<F, T>(list: &Option<RefCell<PetitionFactors>>, map: F) -> Option<T>
+    where
+        F: Fn(&PetitionFactors) -> T,
+    {
+        list.as_ref().map(|refcell| map(&refcell.borrow()))
+    }
+
+    fn on_list<F, R>(&self, kind: FactorListKind, r#do: &F) -> Option<R>
+    where
+        F: Fn(&PetitionFactors) -> R,
+    {
+        match kind {
+            FactorListKind::Threshold => Self::with_list(&self.threshold_factors, r#do),
+            FactorListKind::Override => Self::with_list(&self.override_factors, r#do),
+        }
+    }
+
+    fn both<F, C, T, R>(&self, r#do: F, combine: C) -> R
+    where
+        F: Fn(&PetitionFactors) -> T,
+        C: Fn(Option<T>, Option<T>) -> R,
+    {
+        let t = self.on_list(FactorListKind::Threshold, &r#do);
+        let o = self.on_list(FactorListKind::Override, &r#do);
+        combine(t, o)
+    }
+
+    fn both_void<F, R>(&self, r#do: F)
```

Comment:
yes all are (potentially) mutating, using interior mutability (`RefCell`). But If we prefix all with`possibly_mut` then we get quite long names - e.g. with recent change https://github.com/radixdlt/sargon-mfa/pull/1#discussion_r1736336086 to `map_list_then_form_union` -> `possibly_mut_map_list_then_form_union`.

Reply:
yes all are (potentially) mutating, using interior mutability (`RefCell`). But If we prefix all with`possibly_mut` then we get quite long names - e.g. with recent change https://github.com/radixdlt/sargon-mfa/pull/1#discussion_r1736336086 to `map_list_then_form_union` -> `possibly_mut_map_list_then_form_union`.

Reply:
As I said in https://github.com/radixdlt/sargon-mfa/pull/1#discussion_r1736336086 these are non public methods so that we can minimize a lot of code duplication.